src/simulator/views/map/display/graph_map_display.py

```diff
@@ -1,50 +1,218 @@
-from typing import TYPE_CHECKING
+from typing import List, Dict, TYPE_CHECKING
 
 from algorithms.configuration.maps.map import Map
 from simulator.services.services import Services
 from simulator.views.map.display.map_display import MapDisplay
-from structures import Colour, RED
+from structures import DynamicColour, Colour, RED
+from structures.tracked import Tracked
+from algorithms.classic.sample_based.core.graph import TrackedGraph, Graph, Vertex
 
-if TYPE_CHECKING:
-    from algorithms.classic.sample_based.rrt import Graph, Vertex
+from panda3d.core import NodePath
 
 
 class GraphMapDisplay(MapDisplay):
-    __graph: 'Graph'
+    __graph: Graph
 
-    edge_colour: Colour
-    node_colour: Colour
+    edge_colour: DynamicColour
+    node_colour: DynamicColour
+    __deduced_edge_colour: Colour
+    __deduced_node_colour: Colour
+    __first_frame: bool
 
-    def __init__(self, services: Services, graph: 'Graph', custom_map: Map = None) -> None:
+    def __init__(self, services: Services, graph: Graph, custom_map: Map = None) -> None:
         super().__init__(services, z_index=250, custom_map=custom_map)
         self.__graph = graph
 
         self.edge_colour = self._services.state.add_colour("graph edge", RED)
         self.node_colour = self._services.state.add_colour("graph node", RED)
 
-    def render(self) -> bool:
-        if not super().render():
-            return False
+        self.__deduced_edge_colour = self.edge_colour()
+        self.__deduced_node_colour = self.node_colour()
 
-        self.__graph.walk_dfs(lambda child: self.__render(child))
+        self.__first_frame = True
+        self.__graph_with_irremovable_edges_np = None
 
-        return True
+    def render(self, refresh: bool) -> None:
+        c = self.edge_colour()
+        refresh = refresh or c != self.__deduced_edge_colour
+        self.__deduced_edge_colour = c
+
+        c = self.node_colour()
+        refresh = refresh or c != self.__deduced_node_colour
+        self.__deduced_node_colour = c
+
+        if isinstance(self.__graph, TrackedGraph):
+            self.__render_lazy(refresh)
+        else:
+            self.__render_eager()
+
+    def __render_lazy(self, refresh: bool) -> None:        
+        if self.__graph.edges_removable:
+            self.__render_lazy_edges_removable(refresh)
+        else:
+            self.__render_lazy_edges_irremovable(refresh)
+
+    def __render_lazy_edges_irremovable(self, refresh: bool) -> None:
+        rv = self.get_renderer_view()
+
+        if refresh:
+            def remove_child(current) -> bool:
```

Comment:
defining functions in other functions like this is probably not the way to go code-style wise. especially as there's another function w the same name that does *almost the same thing in the next function below